File: marine_control/src/control_server.cpp
// Copyright 2026 Center for Coastal and Ocean Mapping & NOAA-UNH Joint
// Hydrographic Center, University of New Hampshire
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "marine_control/control_server.hpp"

#include <algorithm>
#include <functional>
#include <stdexcept>
#include <string>
#include <vector>

#include "rclcpp/create_publisher.hpp"
#include "rclcpp/create_subscription.hpp"
#include "rclcpp/create_timer.hpp"

namespace marine_control
{

using marine_control_interfaces::msg::ControlItem;
using marine_control_interfaces::msg::ControlSet;
using marine_control_interfaces::msg::ControlValue;

namespace
{
bool parse_bool(const std::string & s, bool & out)
{
  std::string v = s;
  std::transform(v.begin(), v.end(), v.begin(), [](unsigned char c) {return std::tolower(c);});
  if (v == "true" || v == "1") {out = true; return true;}
  if (v == "false" || v == "0") {out = false; return true;}
  return false;
}
}  // namespace

ControlServer::ControlServer(
  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
  rclcpp::node_interfaces::NodeParametersInterface::SharedPtr node_parameters,
  rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
  rclcpp::node_interfaces::NodeTimersInterface::SharedPtr node_timers,
  rclcpp::node_interfaces::NodeClockInterface::SharedPtr node_clock,
  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
  const ControlServerOptions & options)
: node_parameters_(node_parameters),
  node_clock_(node_clock),
  node_logging_(node_logging),
  options_(options)
{
  if (options_.device_name.empty()) {
    options_.device_name = node_base->get_name();
  }

  // D5: state is RELIABLE + VOLATILE with a heartbeat. Never TRANSIENT_LOCAL —
  // a late joiner gets fresh state from the next heartbeat, not a stale latched
  // sample across the bridge.
  rclcpp::QoS state_qos(rclcpp::KeepLast(1));
  state_qos.reliable();
  state_qos.durability_volatile();
  state_pub_ = rclcpp::create_publisher<ControlSet>(
    node_parameters, node_topics, options_.state_topic, state_qos);

  // change: reliable delivery of the request; fire-and-forget at the app level
  // (success is observed via the next state echo, D1). A small queue absorbs
  // bursts of operator edits.
  rclcpp::QoS change_qos(rclcpp::KeepLast(10));
  change_qos.reliable();
  change_qos.durability_volatile();
  change_sub_ = rclcpp::create_subscription<ControlValue>(
    node_parameters, node_topics, options_.change_topic, change_qos,
    std::bind(&ControlServer::on_change, this, std::placeholders::_1));

  const double period = options_.heartbeat_period_s > 0.0 ? options_.heartbeat_period_s : 1.0;
  heartbeat_timer_ = rclcpp::create_timer(
    node_base, node_timers, node_clock->get_clock(),
    rclcpp::Duration::from_seconds(period),
    std::bind(&ControlServer::publish_state, this));
}

void ControlServer::bind_parameter(
  const std::string & name, const std::string & units, const std::string & group)
{
  if (!node_parameters_->has_parameter(name)) {
    RCLCPP_WARN(
      node_logging_->get_logger(),
      "marine_control: bind_parameter('%s') ignored — parameter not declared. "
      "Declare it (with a descriptor) before binding.", name.c_str());
    return;
  }
  bindings_[name] = Binding{units, group};
}

bool ControlServer::make_item(
  const std::string & name, const Binding & binding, ControlItem & item) const
{
  if (!node_parameters_->has_parameter(name)) {
    return false;
  }
  const auto descriptors = node_parameters_->describe_parameters({name});
  const auto values = node_parameters_->get_parameters({name});
  if (descriptors.empty() || values.empty()) {
    return false;
  }
  const auto & d = descriptors[0];
  const auto & p = values[0];

  item.name = name;
  item.label = name;  // ParameterDescriptor has no label; fall back to the name.
  item.description = d.description;
  item.read_only = d.read_only;
  item.units = binding.units;
  item.group = binding.group;

  switch (p.get_type()) {
    case rclcpp::ParameterType::PARAMETER_DOUBLE:
      item.type = ControlItem::TYPE_FLOAT;
      item.value = std::to_string(p.as_double());
      if (!d.floating_point_range.empty()) {
        item.min_value = d.floating_point_range[0].from_value;
        item.max_value = d.floating_point_range[0].to_value;
        item.step = d.floating_point_range[0].step;
      }
      break;
    case rclcpp::ParameterType::PARAMETER_INTEGER:
      item.type = ControlItem::TYPE_INT;
      item.value = std::to_string(p.as_int());
      if (!d.integer_range.empty()) {
        item.min_value = static_cast<double>(d.integer_range[0].from_value);
        item.max_value = static_cast<double>(d.integer_range[0].to_value);
        item.step = static_cast<double>(d.integer_range[0].step);
      }
      break;
    case rclcpp::ParameterType::PARAMETER_BOOL:
      item.type = ControlItem::TYPE_BOOL;
      item.value = p.as_bool() ? "true" : "false";
      break;
    case rclcpp::ParameterType::PARAMETER_STRING:
      item.type = ControlItem::TYPE_STRING;
      item.value = p.as_string();
      break;
    default:
      // Arrays and unset types are not (yet) renderable as a single control.
      RCLCPP_WARN_ONCE(
        node_logging_->get_logger(),
        "marine_control: parameter '%s' has an unsupported type for a control; skipping.",
        name.c_str());
      return false;
  }
  return true;
}

ControlSet ControlServer::build_control_set() const
{
  ControlSet set;
  set.header.stamp = node_clock_->get_clock()->now();
  set.device_name = options_.device_name;
  for (const auto & [name, binding] : bindings_) {
    ControlItem item;
    if (make_item(name, binding, item)) {
      set.items.push_back(item);
    }
  }
  return set;
}

void ControlServer::publish_state()
{
  if (state_pub_) {
    state_pub_->publish(build_control_set());
  }
}

void ControlServer::on_change(const ControlValue::SharedPtr msg)
{
  const std::string & name = msg->name;
  if (bindings_.find(name) == bindings_.end() || !node_parameters_->has_parameter(name)) {
    RCLCPP_WARN(
      node_logging_->get_logger(),
      "marine_control: change for unbound/unknown control '%s' ignored.", name.c_str());
    return;
  }

  const auto values = node_parameters_->get_parameters({name});
  if (values.empty()) {
    return;
  }

  rclcpp::Parameter new_param;
  try {
    switch (values[0].get_type()) {
      case rclcpp::ParameterType::PARAMETER_DOUBLE:
        new_param = rclcpp::Parameter(name, std::stod(msg->value));
        break;
      case rclcpp::ParameterType::PARAMETER_INTEGER:
        new_param = rclcpp::Parameter(name, static_cast<int64_t>(std::stoll(msg->value)));
        break;
      case rclcpp::ParameterType::PARAMETER_BOOL: {
          bool b = false;
          if (!parse_bool(msg->value, b)) {
            throw std::invalid_argument("expected true/false");
          }
          new_param = rclcpp::Parameter(name, b);
          break;
        }
      case rclcpp::ParameterType::PARAMETER_STRING:
        new_param = rclcpp::Parameter(name, msg->value);
        break;
      default:
        RCLCPP_WARN(
          node_logging_->get_logger(),
          "marine_control: control '%s' has an unsupported type; change ignored.", name.c_str());
        return;
    }
  } catch (const std::exception & e) {
    RCLCPP_WARN(
      node_logging_->get_logger(),
      "marine_control: could not parse change value '%s' for '%s': %s",
      msg->value.c_str(), name.c_str(), e.what());
    return;
  }

  // The node's own on-set validation runs here and may reject the value.
  const auto results = node_parameters_->set_parameters({new_param});
  if (!results.empty() && !results[0].successful) {
    RCLCPP_WARN(
      node_logging_->get_logger(),
      "marine_control: change to '%s' rejected: %s", name.c_str(), results[0].reason.c_str());
  }
  // Echo current state either way (D1): the operator sees the applied value, or
  // the unchanged value if it was rejected.
  publish_state();
}

}  // namespace marine_control

File: marine_control_py/marine_control_py/control_server.py
# Copyright 2026 Center for Coastal and Ocean Mapping & NOAA-UNH Joint
# Hydrographic Center, University of New Hampshire
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Boat-side device-control server (Python sibling of marine_control::ControlServer).

Mirrors the C++ contract (ADR-0003 D4/D5/D6): a node declares parameters with
descriptors, binds them with bind_parameter(), and this server

  - publishes a marine_control_interfaces/ControlSet (state, device -> operator)
    on a periodic heartbeat with the bridge-correct QoS (D5: RELIABLE + VOLATILE,
    never TRANSIENT_LOCAL), and
  - applies inbound marine_control_interfaces/ControlValue (change,
    operator -> device) by setting the bound parameter -- whose own on-set
    validation the node already owns -- then re-publishes state so the echo
    confirms the change (D1, fire-and-forget).

No Qt: the operator station renders the published ControlSet.

Threading contract: the heartbeat timer and the change subscription share a
dedicated mutually-exclusive callback group, so they never run concurrently with
each other even under a multi-threaded executor. Call bind_parameter() during
setup, before the node starts spinning -- it mutates the binding table without
locking and is not safe to call concurrently with the running callbacks.
Likewise, drop the server only when the node is not spinning; the timer/sub are
torn down with it, and an in-flight callback on another thread would race.

Step note: a parameter declared with a non-zero descriptor step makes rclpy
reject in-range-but-off-step changes (e.g. 0.63 against step 0.05) rather than
snapping them. Declare step=0.0 for controls that should accept any in-range
value, or have the operator UI snap to the step before sending.

Lifecycle note: the publisher, subscription, and timer are created on
construction, so the server is active as soon as it is constructed. Construct it
when you want it active. A safety-critical adopter (ADR-0003 D8.3) that needs
confirmation/audit beyond fire-and-forget layers that on top of this mechanism.
"""

import copy
from dataclasses import dataclass

from marine_control_interfaces.msg import ControlItem, ControlSet, ControlValue

from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
from rclpy.parameter import Parameter
from rclpy.qos import DurabilityPolicy, HistoryPolicy, QoSProfile, ReliabilityPolicy


@dataclass
class ControlServerOptions:
    """Configuration for a ControlServer (mirrors the C++ ControlServerOptions)."""

    # UI title for the panel. Empty -> the node name is used.
    device_name: str = ''
    # state (device->operator) and change (operator->device) topics. Relative
    # names resolve under the node namespace; the platform bridge config wires
    # state boat->operator and change operator->boat (ADR-0003 D7).
    state_topic: str = '~/control/state'
    change_topic: str = '~/control/change'
    # Heartbeat republish period (s) for state. A late joiner gets fresh state
    # from the next heartbeat rather than a stale latched sample (D5).
    heartbeat_period_s: float = 1.0


def _parse_bool(text):
    """Parse a bool from true/false/1/0 (case-insensitive); raise ValueError otherwise."""
    value = text.strip().lower()
    if value in ('true', '1'):
        return True
    if value in ('false', '0'):
        return False
    raise ValueError("expected true/false, got '{}'".format(text))


class ControlServer:
    """Publishes a node's bound parameters as a ControlSet and applies changes."""

    def __init__(self, node, options=None):
        """Bind to an rclpy node (Node or LifecycleNode); it must outlive the server."""
        self._node = node
        # Copy so resolving device_name below never mutates a caller-shared
        # options instance (the C++ sibling takes options by value).
        self._options = copy.copy(options) if options is not None else ControlServerOptions()
        if not self._options.device_name:
            self._options.device_name = node.get_name()
        self._logger = node.get_logger()
        # name -> (units, group). Mutated only by bind_parameter (see contract).
        self._bindings = {}
        # Names already warned about as unsupported, so build_control_set does
        # not spam the log once per heartbeat (mirrors C++ RCLCPP_WARN_ONCE).
        self._unsupported_warned = set()

        # The heartbeat timer (publish_state) and the change subscription
        # (on_change) both read _bindings and touch parameters; a single
        # mutually-exclusive group serializes them regardless of the executor.
        self._callback_group = MutuallyExclusiveCallbackGroup()

        # D5: state is RELIABLE + VOLATILE with a heartbeat. Never
        # TRANSIENT_LOCAL -- a late joiner gets fresh state from the next
        # heartbeat, not a stale latched sample across the bridge.
        state_qos = QoSProfile(
            depth=1,
            history=HistoryPolicy.KEEP_LAST,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE)
        self._state_pub = node.create_publisher(
            ControlSet, self._options.state_topic, state_qos)

        # change: reliable delivery of the request; fire-and-forget at the app
        # level (success is observed via the next state echo, D1).
        change_qos = QoSProfile(
            depth=10,
            history=HistoryPolicy.KEEP_LAST,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE)
        self._change_sub = node.create_subscription(
            ControlValue, self._options.change_topic, self._on_change, change_qos,
            callback_group=self._callback_group)

        period = self._options.heartbeat_period_s
        if period <= 0.0:
            period = 1.0
        self._timer = node.create_timer(
            period, self.publish_state, callback_group=self._callback_group)

    def bind_parameter(self, name, units='', group=''):
        """Expose an already-declared parameter as a control (descriptor populates the item)."""
        if not self._node.has_parameter(name):
            self._logger.warn(
                "marine_control: bind_parameter('{}') ignored -- parameter not "
                'declared. Declare it (with a descriptor) before binding.'.format(name))
            return
        self._bindings[name] = (units, group)

    def build_control_set(self):
        """Assemble the current ControlSet from the bound parameters (stamp = now)."""
        msg = ControlSet()
        msg.header.stamp = self._node.get_clock().now().to_msg()
        msg.device_name = self._options.device_name
        for name in sorted(self._bindings):
            item = self._make_item(name)
            if item is not None:
                msg.items.append(item)
        return msg

    def publish_state(self):
        """Publish the current ControlSet immediately (also called on the heartbeat)."""
        self._state_pub.publish(self.build_control_set())

    def _make_item(self, name):
        """Translate one declared parameter into a ControlItem, or None if not renderable."""
        if not self._node.has_parameter(name):
            return None
        descriptor = self._node.describe_parameter(name)
        param = self._node.get_parameter(name)
        units, group = self._bindings[name]

        item = ControlItem()
        item.name = name
        item.label = name  # ParameterDescriptor has no label; fall back to the name.
        item.description = descriptor.description
        item.read_only = descriptor.read_only
        item.units = units
        item.group = group

        ptype = param.type_
        if ptype == Parameter.Type.DOUBLE:
            item.type = ControlItem.TYPE_FLOAT
            item.value = repr(float(param.value))  # shortest round-trippable
            if descriptor.floating_point_range:
                rng = descriptor.floating_point_range[0]
                item.min_value = rng.from_value
                item.max_value = rng.to_value
                item.step = rng.step
        elif ptype == Parameter.Type.INTEGER:
            item.type = ControlItem.TYPE_INT
            item.value = str(int(param.value))
            if descriptor.integer_range:
                rng = descriptor.integer_range[0]
                item.min_value = float(rng.from_value)
                item.max_value = float(rng.to_value)
                item.step = float(rng.step)
        elif ptype == Parameter.Type.BOOL:
            item.type = ControlItem.TYPE_BOOL
            item.value = 'true' if param.value else 'false'
        elif ptype == Parameter.Type.STRING:
            item.type = ControlItem.TYPE_STRING
            item.value = param.value
        else:
            # Arrays and unset types are not (yet) renderable as a single
            # control. Warn once per name -- build_control_set runs every
            # heartbeat, so an unconditional warn would spam the log.
            if name not in self._unsupported_warned:
                self._unsupported_warned.add(name)
                self._logger.warn(
                    "marine_control: parameter '{}' has an unsupported type for "
                    'a control; skipping.'.format(name))
            return None
        return item

    def _on_change(self, msg):
        """Apply one ControlValue to the bound parameter, then echo state (D1)."""
        name = msg.name
        if name not in self._bindings or not self._node.has_parameter(name):
            self._logger.warn(
                "marine_control: change for unbound/unknown control '{}' "
                'ignored.'.format(name))
            return

        current = self._node.get_parameter(name)
        old_value = str(current.value)
        ptype = current.type_

        # Python int()/float() already reject trailing garbage ("0.75abc") and
        # integer truncation ("3.5") by raising ValueError -- the parse hardening
        # the C++ sibling needed std::from_chars for is free here.
        try:
            if ptype == Parameter.Type.DOUBLE:
                new_param = Parameter(name, Parameter.Type.DOUBLE, float(msg.value.strip()))
            elif ptype == Parameter.Type.INTEGER:
                new_param = Parameter(name, Parameter.Type.INTEGER, int(msg.value.strip()))
            elif ptype == Parameter.Type.BOOL:
                new_param = Parameter(name, Parameter.Type.BOOL, _parse_bool(msg.value))
            elif ptype == Parameter.Type.STRING:
                new_param = Parameter(name, Parameter.Type.STRING, msg.value)
            else:
                self._logger.warn(
                    "marine_control: control '{}' has an unsupported type; "
                    'change ignored.'.format(name))
                return
        except ValueError:
            self._logger.warn(
                "marine_control: could not parse change value '{}' for '{}' "
                '(expected {}); ignored.'.format(msg.value, name, ptype.name))
            return

        # The node's own on-set validation runs here and may reject the value.
        results = self._node.set_parameters([new_param])
        if not results or not results[0].successful:
            reason = results[0].reason if results else 'no result returned'
            self._logger.warn(
                "marine_control: change to '{}' rejected: {}".format(name, reason))
        else:
            # Audit trail: every applied change is logged old -> new. Groundwork
            # for the e-stop adopter's stricter change-audit (ADR-0003 D8.3).
            self._logger.info(
                "marine_control: '{}' changed {} -> {}".format(
                    name, old_value, new_param.value))
        # Echo current state either way (D1): the operator sees the applied
        # value, or the unchanged value if it was rejected.
        self.publish_state()
